### ISSUE TITLE
Allow non ordered tree list

### ISSUE THREAD

**Issue by jmatsushita** [reactabular/reactabular#195]
Hi there,

Great component, thanks for your work! 

I just bumped into a problem with `reactabular-tree`. Inserting a node between a parent and its children confuses the tree rendering.

```
const rows = [
  {
    id: 100,
    name: 'John',
    country: 'fi'
  },
  {
    id: 150,
    name: 'Alice',
    country: 'fi',
  },
  {
    id: 200,
    name: 'Mary',
    country: 'fi',
    parent: 100
  }
]
```

![image](https://cloud.githubusercontent.com/assets/356097/18356749/4fe03c9c-75ee-11e6-94ea-b36aec643c0a.png)

**Comment by bebraw** [comment]
Yeah, it's very peculiar about the structure.

The question is, how should it work in this case?

Maybe it would make sense to implement a little function that fixes the data structure (fixes the ordering). That way we wouldn't have to touch the current algorithms and we would get nice render out of it.

**Comment by jmatsushita** [comment]
I see your point if the array order is meaningful in other situations, then it could be confusing if trees don't behave the same way. On the other hand, you could argue that if you want a tree, then you care more about the data structure and would want the library to figure out what to do with the order.

So my vote would go to do the ordering fixing behind the scenes and by default for trees.

But talking about little functions, it would be quite nice to also allow taking directly a nested object, in the way that d3 uses [a hierarchy](https://github.com/d3/d3-hierarchy#hierarchy) (i.e. a nested object) by default and has a utility function [stratify](https://github.com/d3/d3-hierarchy#stratify) to create that nested format from a tabular format with a parent column.

Given that you `pack` `unpack` seem to deal with that, maybe getting closer to this API would make things more flexible (I know that I found it missing just now :) )

**Comment by bebraw** [comment]
> So my vote would go to do the ordering fixing behind the scenes and by default for trees.

There's an additional option. I could run a validator in development and give a warning.

I can provide a fixing function, but we need to be a little careful with that to avoid hurting performance too much. So we would run the fix when loading data and avoid it during queries.

> But talking about little functions, it would be quite nice to also allow taking directly a nested object, in the way that d3 uses a hierarchy (i.e. a nested object) by default and has a utility function stratify to create that nested format from a tabular format with a parent column.

Nested structures can be converted to a structure the algorithms expect. Flat structure is convenient when it comes to rendering a table. So it's a similar problem like validation or fixing, another utility function or two.

Maybe a starting point would be to start from that d3 structure and write a transformation for that.

> Given that you pack unpack seem to deal with that, maybe getting closer to this API would make things more flexible (I know that I found it missing just now :) )

I had to implement `pack` and `unpack` to deal with filtering and sorting. For instance I specifically want to sort based on root nodes while skipping children, hence packing.

Filtering operates on root level as well, but a smarter algorithm could be implemented. Perhaps it should search the whole tree and match against visible children as well.

--

To recap, I think we'll want at least the following utility functions:

* `validate` - Accepts a tree structure and raises an error if it's not valid.
* `fix` - Accepts a tree structure and tries to fix it. Start from the structure you suggested.
* `nestedToFlat` (not sure of the name) - Accepts a nested structure (d3 style) and converts it to a flat one (children contain parent relations, no gaps).

I wouldn't mind a PR for these. Note that you would need to implement tests. Currently the tree functionality isn't tested, but adding a `tests` directory and following the existing conventions should work without any hassle. The global testing setup picks that up.

**Comment by jmatsushita** [comment]
Thanks for the recap. It all sounds good. I don't have much time to do a PR unfortunately. Sorry!

**Comment by bebraw** [comment]
No worries. I might tackle this one day unless someone else gets first.

**Comment by bebraw** [comment]
There's `tree.flatten` now (`nestedToFlat`). I also added some API docs.

I'm thinking the API could use a small rewrite to make the naming consistent. A couple of functions need to change slightly.

**Comment by bebraw** [comment]
Hi,

I pushed the tree logic to [treetabular](https://github.com/reactabular/treetabular). Can you check the project and open issues based on your needs? The API has grown quite a bit since the last time you looked I think.
